File: cmd/nomad-p2p/bpf.go
package main

import (
	"fmt"
	"log"
	"net"
	"os/exec"
	"strings"

	"github.com/cilium/ebpf"
	"github.com/cilium/ebpf/link"
)

type BPFMaps struct {
	ContainerRouteMap *ebpf.Map
	NodeDynamicMap    *ebpf.Map
	RouteMissRingbuf  *ebpf.Map
	GeneveIfindexMap  *ebpf.Map
	TunnelCfgMap      *ebpf.Map
	VIPMap            *ebpf.Map
	VIPStatsMap       *ebpf.Map
	ACLMap            *ebpf.Map
	PortACLMap        *ebpf.Map
	DefaultPolicy     *ebpf.Map
}

type NodeEndpoint struct {
	PublicIP   uint32
	PublicPort uint16
	Pad        uint16
}

type VIPInfo struct {
	Backends [16]uint32
	Count    uint8
	Pad      [3]byte
	NextIdx  uint32
}

type TunnelCfg struct {
	TunnelID uint32
}

func (a *Agent) loadBPF() error {
	spec, err := ebpf.LoadCollectionSpec("bin/mesh.bpf.o")
	if err != nil {
		return fmt.Errorf("load mesh spec: %w", err)
	}
	coll, err := ebpf.NewCollection(spec)
	if err != nil {
		return fmt.Errorf("load mesh collection: %w", err)
	}
	a.maps = &BPFMaps{
		ContainerRouteMap: coll.Maps["CONTAINER_ROUTE_MAP"],
		NodeDynamicMap:    coll.Maps["NODE_DYNAMIC_MAP"],
		RouteMissRingbuf:  coll.Maps["ROUTE_MISS_RINGBUF"],
		GeneveIfindexMap:  coll.Maps["GENEVE_IFINDEX_MAP"],
		TunnelCfgMap:      coll.Maps["TUNNEL_CFG_MAP"],
	}
	a.meshColl = coll

	pinDir := "/sys/fs/bpf/nomad-p2p"
	exec.Command("mkdir", "-p", pinDir).Run()
	if a.maps.ContainerRouteMap != nil {
		a.maps.ContainerRouteMap.Pin(pinDir + "/container_route")
	}
	if a.maps.NodeDynamicMap != nil {
		a.maps.NodeDynamicMap.Pin(pinDir + "/node_dynamic")
	}
	if prog, ok := a.meshColl.Programs["egress_p2p_mesh"]; ok {
		prog.Pin(pinDir + "/mesh_prog")
		log.Printf("[agent] pinned egress_p2p_mesh program")
	}
	log.Printf("[agent] mesh BPF maps pinned to %s", pinDir)

	if a.maps.TunnelCfgMap != nil {
		key := uint32(0)
		cfg := TunnelCfg{TunnelID: uint32(a.cfg.TunnelVNI)}
		a.maps.TunnelCfgMap.Update(key, cfg, 0)
	}

	if a.cfg.VIPEnabled {
		vipSpec, err := ebpf.LoadCollectionSpec("bin/vip_balancer.bpf.o")
		if err != nil {
			log.Printf("[agent] VIP balancer BPF not found: %v", err)
		} else {
			vipColl, err := ebpf.NewCollection(vipSpec)
			if err != nil {
				log.Printf("[agent] VIP balancer load failed: %v", err)
			} else {
				a.maps.VIPMap = vipColl.Maps["VIP_MAP"]
				a.maps.VIPStatsMap = vipColl.Maps["VIP_STATS_MAP"]
				a.vipColl = vipColl
				log.Printf("[agent] VIP balancer BPF loaded")
			}
		}
	}

	fwSpec, err := ebpf.LoadCollectionSpec("bin/firewall.bpf.o")
	if err != nil {
		log.Printf("[agent] firewall BPF not found: %v", err)
	} else {
		fwColl, err := ebpf.NewCollection(fwSpec)
		if err != nil {
			log.Printf("[agent] firewall load failed: %v", err)
		} else {
			a.maps.ACLMap = fwColl.Maps["ACL_MAP"]
			a.maps.PortACLMap = fwColl.Maps["PORT_ACL_MAP"]
			a.maps.DefaultPolicy = fwColl.Maps["DEFAULT_POLICY"]
			a.fwColl = fwColl
			a.loadACLsFromConfig()
		}
	}

	return nil
}

func (a *Agent) attachBPF() error {
	iface, err := findDefaultInterface()
	if err != nil {
		log.Printf("[agent] no network interface found for BPF attachment: %v", err)
		return nil
	}
	ifIndex := uint32(iface.Index)
	log.Printf("[agent] attaching BPF to interface %s (index %d)", iface.Name, iface.Index)

	if a.meshColl != nil {
		if xdpProg, ok := a.meshColl.Programs["xdp_pass"]; ok {
			l, err := link.AttachXDP(link.XDPOptions{
				Interface: iface.Index,
				Program:   xdpProg,
			})
			if err != nil {
				log.Printf("[agent] XDP attach failed: %v", err)
			} else {
				a.bpfLinks = append(a.bpfLinks, l)
				log.Printf("[agent] XDP attached to %s", iface.Name)
			}
		}

		if tcProg, ok := a.meshColl.Programs["egress_p2p_mesh"]; ok {
			if err := a.attachTC(ifIndex, tcProg, "egress"); err != nil {
				log.Printf("[agent] TC egress attach failed: %v", err)
			} else {
				log.Printf("[agent] TC egress attached to %s", iface.Name)
			}
		}
	}

	if a.fwColl != nil {
		if fwProg, ok := a.fwColl.Programs["tc_ingress_firewall"]; ok {
			if err := a.attachTC(ifIndex, fwProg, "ingress"); err != nil {
				log.Printf("[agent] TC ingress attach failed: %v", err)
			} else {
				log.Printf("[agent] TC ingress firewall attached to %s", iface.Name)
			}
		}
	}

	if a.vipColl != nil {
		if vipProg, ok := a.vipColl.Programs["vip_load_balance"]; ok {
			if err := a.attachCgroup(vipProg); err != nil {
				log.Printf("[agent] cgroup attach failed: %v", err)
			} else {
				log.Printf("[agent] VIP cgroup program attached")
			}
		}
	}

	return nil
}

func (a *Agent) attachTC(ifIndex uint32, prog *ebpf.Program, direction string) error {
	iface, err := net.InterfaceByIndex(int(ifIndex))
	if err != nil {
		return fmt.Errorf("get interface: %w", err)
	}

	cmd := exec.Command("tc", "qdisc", "add", "dev", iface.Name, "clsact")
	cmd.Run()

	pinPath := fmt.Sprintf("/sys/fs/bpf/tc/%s_%d", direction, ifIndex)
	exec.Command("rm", "-f", pinPath).Run()
	if err := exec.Command("bpftool", "prog", "pin", "id",
		fmt.Sprintf("%d", prog.FD()), pinPath).Run(); err != nil {
		return fmt.Errorf("pin prog: %w", err)
	}

	cmd = exec.Command("tc", "filter", "add", "dev", iface.Name, direction,
		"pref", "1", "handle", "1", "bpf", "direct-action", "pinned", pinPath)
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("tc filter add: %w: %s", err, out)
	}

	return nil
}

func (a *Agent) attachCgroup(prog *ebpf.Program) error {
	l, err := link.AttachCgroup(link.CgroupOptions{
		Path:    "/sys/fs/cgroup",
		Program: prog,
		Attach:  ebpf.AttachCGroupInet4Connect,
	})
	if err != nil {
		return fmt.Errorf("attach cgroup: %w", err)
	}
	a.bpfLinks = append(a.bpfLinks, l)
	return nil
}

func findDefaultInterface() (*net.Interface, error) {
	out, err := exec.Command("ip", "route", "show", "default").CombinedOutput()
	if err == nil {
		fields := strings.Fields(string(out))
		for i, f := range fields {
			if f == "dev" && i+1 < len(fields) {
				return net.InterfaceByName(fields[i+1])
			}
		}
	}

	for _, name := range []string{"eth0", "ens3", "wlan0", "enp0s3"} {
		iface, err := net.InterfaceByName(name)
		if err == nil {
			return iface, nil
		}
	}

	ifaces, err := net.Interfaces()
	if err != nil {
		return nil, err
	}
	for _, iface := range ifaces {
		if iface.Flags&net.FlagUp != 0 && iface.Flags&net.FlagLoopback == 0 {
			return &iface, nil
		}
	}
	return nil, fmt.Errorf("no suitable network interface found")
}

func (a *Agent) updateGeneveIfindex() {
	if a.maps.GeneveIfindexMap == nil {
		return
	}
	iface, err := net.InterfaceByName(a.cfg.TunnelDevice)
	if err != nil {
		log.Printf("[agent] geneve device %s not found for ifindex update", a.cfg.TunnelDevice)
		return
	}
	key := uint32(0)
	val := uint32(iface.Index)
	a.maps.GeneveIfindexMap.Update(key, val, 0)
	log.Printf("[agent] geneve ifindex %d stored in BPF map", iface.Index)
}

File: agent/src/bpf.rs
use anyhow::{Context, Result};
use aya::{
    programs::{CgroupAttachMode, SchedClassifier, TcAttachType, Xdp, XdpFlags},
    Ebpf, EbpfLoader, MapData, maps::HashMap,
};
use nomad_p2p_common::{AclRule, PeerEndpoint, TunnelCfg, VipInfo};
use std::path::Path;

pub struct BpfMaps {
    pub container_route: Option<HashMap<MapData, u32, u32>>,
    pub node_dynamic: Option<HashMap<MapData, u32, PeerEndpoint>>,
    pub route_miss_ringbuf: Option<aya::maps::RingBuf<MapData>>,
    pub geneve_ifindex: Option<HashMap<MapData, u32, u32>>,
    pub tunnel_cfg: Option<HashMap<MapData, u32, TunnelCfg>>,
    pub vip_map: Option<HashMap<MapData, u32, VipInfo>>,
    pub acl_map: Option<HashMap<MapData, u32, AclRule>>,
    pub default_policy: Option<HashMap<MapData, u32, u8>>,
}

pub struct BpfManager {
    pub mesh: Ebpf,
    pub fw: Option<Ebpf>,
    pub vip: Option<Ebpf>,
    pub maps: BpfMaps,
    pub links: Vec<Box<dyn aya::programs::Link>>,
    pub pinned: bool,
}

impl BpfManager {
    pub fn load() -> Result<Self> {
        let mut mesh = EbpfLoader::new()
            .load_file(Path::new("bin/mesh.bpf.o"))
            .context("load mesh BPF")?;
        aya_log::EbpfLogger::init(&mut mesh).ok();

        let maps = BpfMaps {
            container_route: mesh.take_map("CONTAINER_ROUTE_MAP").ok().and_then(|d| HashMap::new(d, 0).ok()),
            node_dynamic: mesh.take_map("NODE_DYNAMIC_MAP").ok().and_then(|d| HashMap::new(d, 0).ok()),
            route_miss_ringbuf: mesh.take_map("ROUTE_MISS_RINGBUF").ok().and_then(|d| aya::maps::RingBuf::new(d).ok()),
            geneve_ifindex: mesh.take_map("GENEVE_IFINDEX_MAP").ok().and_then(|d| HashMap::new(d, 0).ok()),
            tunnel_cfg: mesh.take_map("TUNNEL_CFG_MAP").ok().and_then(|d| HashMap::new(d, 0).ok()),
            vip_map: None,
            acl_map: None,
            default_policy: None,
        };

        let fw = match EbpfLoader::new().load_file(Path::new("bin/firewall.bpf.o")) {
            Ok(mut fw) => {
                aya_log::EbpfLogger::init(&mut fw).ok();
                Some(fw)
            }
            Err(e) => {
                tracing::warn!("firewall BPF not loaded: {}", e);
                None
            }
        };

        let vip = match EbpfLoader::new().load_file(Path::new("bin/vip_balancer.bpf.o")) {
            Ok(mut vip) => {
                aya_log::EbpfLogger::init(&mut vip).ok();
                Some(vip)
            }
            Err(e) => {
                tracing::warn!("VIP BPF not loaded: {}", e);
                None
            }
        };

        Ok(Self {
            mesh,
            fw,
            vip,
            maps,
            links: vec![],
            pinned: false,
        })
    }

    pub fn set_tunnel_cfg(&self, tunnel_id: u32) -> Result<()> {
        if let Some(ref map) = self.maps.tunnel_cfg {
            map.insert(&0u32, &TunnelCfg { tunnel_id }, 0)?;
        }
        Ok(())
    }

    pub fn set_geneve_ifindex(&self, ifindex: u32) -> Result<()> {
        if let Some(ref map) = self.maps.geneve_ifindex {
            map.insert(&0u32, &ifindex, 0)?;
        }
        Ok(())
    }

    pub fn update_peer(&self, overlay_ip: u32, ep: &PeerEndpoint) -> Result<()> {
        if let Some(ref map) = self.maps.node_dynamic {
            map.insert(&overlay_ip, ep, 0)?;
        }
        Ok(())
    }

    pub fn remove_peer(&self, overlay_ip: u32) -> Result<()> {
        if let Some(ref map) = self.maps.node_dynamic {
            map.remove(&overlay_ip)?;
        }
        Ok(())
    }

    pub fn update_container_route(&self, container_ip: u32, route_vni: u32) -> Result<()> {
        if let Some(ref map) = self.maps.container_route {
            map.insert(&container_ip, &route_vni, 0)?;
        }
        Ok(())
    }

    pub fn remove_container_route(&self, container_ip: u32) -> Result<()> {
        if let Some(ref map) = self.maps.container_route {
            map.remove(&container_ip)?;
        }
        Ok(())
    }

    pub fn attach_all(&mut self, ifindex: u32) -> Result<()> {
        self.attach_xdp(ifindex)?;
        self.attach_tc(ifindex, "egress")?;
        if let Some(ref mut fw) = self.fw {
            self.attach_fw_tc(fw, ifindex)?;
        }
        if let Some(ref mut vip) = self.vip {
            self.attach_cgroup_vip(vip)?;
        }
        Ok(())
    }

    fn attach_xdp(&mut self, ifindex: u32) -> Result<()> {
        let prog = self.mesh.program_mut("xdp_pass").context("xdp_pass not found")?;
        let xdp: &mut Xdp = prog.try_into()?;
        let link = xdp.attach(ifindex, XdpFlags::default())?;
        self.links.push(Box::new(link));
        Ok(())
    }

    fn attach_tc(&mut self, ifindex: u32, _direction: &str) -> Result<()> {
        let prog = self.mesh.program_mut("egress_p2p_mesh").context("egress_p2p_mesh not found")?;
        let tc: &mut SchedClassifier = prog.try_into()?;
        let link = tc.attach(ifindex, TcAttachType::Egress)?;
        self.links.push(Box::new(link));
        Ok(())
    }

    fn attach_fw_tc(&mut self, fw: &mut Ebpf, ifindex: u32) -> Result<()> {
        if let Some(prog) = fw.program_mut("tc_ingress_firewall") {
            let tc: &mut SchedClassifier = prog.try_into()?;
            let link = tc.attach(ifindex, TcAttachType::Ingress)?;
            self.links.push(Box::new(link));
        }
        Ok(())
    }

    fn attach_cgroup_vip(&mut self, vip: &mut Ebpf) -> Result<()> {
        if let Some(prog) = vip.program_mut("vip_load_balance") {
            let cg: &mut aya::programs::CgroupSkb = prog.try_into()?;
            let link = cg.attach(CgroupAttachMode::Single)?;
            self.links.push(Box::new(link));
        }
        Ok(())
    }
}

pub async fn find_default_ifindex() -> Result<u32> {
    let out = tokio::process::Command::new("ip")
        .args(["route", "show", "default"])
        .output()
        .await
        .context("ip route failed")?;
    let stdout = String::from_utf8_lossy(&out.stdout);
    for line in stdout.lines() {
        let parts: Vec<&str> = line.split_whitespace().collect();
        for (i, part) in parts.iter().enumerate() {
            if *part == "dev" && i + 1 < parts.len() {
                let name = parts[i + 1];
                let out = tokio::process::Command::new("ip")
                    .args(["link", "show", name])
                    .output()
                    .await?;
                let stdout = String::from_utf8_lossy(&out.stdout);
                if let Some(line) = stdout.lines().next() {
                    if let Some(idx_str) = line.split(':').next() {
                        if let Ok(idx) = idx_str.trim().parse::<u32>() {
                            return Ok(idx);
                        }
                    }
                }
            }
        }
    }
    Ok(1)
}

pub async fn find_geneve_ifindex(dev: &str) -> Result<u32> {
    let out = tokio::process::Command::new("ip")
        .args(["link", "show", dev])
        .output()
        .await?;
    let stdout = String::from_utf8_lossy(&out.stdout);
    if let Some(line) = stdout.lines().next() {
        if let Some(idx_str) = line.split(':').next() {
            if let Ok(idx) = idx_str.trim().parse::<u32>() {
                return Ok(idx);
            }
        }
    }
    Err(anyhow::anyhow!("Geneve device {} not found", dev))
}
